### PR TITLE
Join By Column Names 

### PR DESCRIPTION
**Pull request by vibhatha**

### CODE REVIEW

**Review by supunkamburugamuve — APPROVED** [review]

File: python/pycylon/data/table.pyx [hunk 1]
```diff
@@ -198,38 +206,107 @@ cdef class Table:
...
+        return isinstance(obj, Table)
+
+    def _resolve_column_indices_from_column_names(self, column_names: List[
+        str], op_column_names: List[str]) -> List[int]:
+        resolve_col_ids = []
+        for op_col_id, op_column_name in enumerate(op_column_names):
+            for col_id, column_name in enumerate(column_names):
+                if op_column_name == column_name:
+                    resolve_col_ids.append(col_id)
+        return resolve_col_ids
```
Comment (nirandaperera) [inline python/pycylon/data/table.pyx:219]: what should we do if the provided column name strings are not found?

Comment (supunkamburugamuve) [reply python/pycylon/data/table.pyx:219]: Let's throw an exception

Comment (vibhatha) [reply python/pycylon/data/table.pyx:219]: @supunkamburugamuve @nirandaperera there is already a check within join and distjoin to verify if there are any elements in the returned list. If no items are there an exception is thrown. 

But I am going to make it neater and pushing it to column retrieval and add a function call.

Comment (vibhatha) [reply python/pycylon/data/table.pyx:219]: @supunkamburugamuve @nirandaperera 
Updated.

File: python/pycylon/data/table.pyx [hunk 2]
```diff
@@ -198,38 +206,110 @@ cdef class Table:
...
+            raise TypeError("kwargs 'on' or 'left_on' and 'right_on' must be provided")
+
+        if not (left_cols and isinstance(left_cols[0], int)) and not (right_cols and isinstance(
+                right_cols[0], int)):
+            raise TypeError("kwargs 'on' or 'left_on' and 'right_on' must be type List and contain "
+                            "int type or str type and cannot be None")
+    def _is_column_indices_viable(self, left_cols, right_cols):
+        return left_cols and right_cols
+
+    def _check_column_names_viable(self, left_cols, right_cols):
```
Comment (vibhatha) [inline python/pycylon/data/table.pyx:262]: @nirandaperera @supunkamburugamuve 
This is the function call for check validity of non-zero values for resolved column indices

File: python/pycylon/data/table.pyx [hunk 3]
```diff
@@ -198,38 +206,110 @@ cdef class Table:
...
+            elif isinstance(left_cols[0], int) and isinstance(right_cols[0], int):
+                return left_cols, right_cols
+        ## Check if Passed values are based on common column names in two tables
+        elif column_names and isinstance(column_names, List):
+            if isinstance(column_names[0], str):
+                left_cols = self._resolve_column_indices_from_column_names(self.column_names,
+                                                                           column_names)
+                right_cols = self._resolve_column_indices_from_column_names(table.column_names,
+                                                                            column_names)
+                self._check_column_names_viable(left_cols, right_cols)
```
Comment (vibhatha) [inline python/pycylon/data/table.pyx:248]: @nirandaperera @supunkamburugamuve 
the check is called before the values are returned to join or distjoin

File: python/pycylon/data/table.pyx [hunk 4]
```diff
@@ -198,38 +206,110 @@ cdef class Table:
...
+        table_col_names_list = [self.column_names, table.column_names]
+
+        if left_cols and right_cols and isinstance(left_cols, List) and isinstance(right_cols,
+                                                                                   List):
+            if isinstance(left_cols[0], str) and isinstance(right_cols[0], str):
+                left_cols = self._resolve_column_indices_from_column_names(self.column_names,
+                                                                           left_cols)
+                right_cols = self._resolve_column_indices_from_column_names(table.column_names,
+                                                                            right_cols)
+                self._check_column_names_viable(left_cols, right_cols)
```
Comment (vibhatha) [inline python/pycylon/data/table.pyx:236]: @nirandaperera @supunkamburugamuve 
the check is called before the values are returned to join or distjoin

**Review by nirandaperera — APPROVED** [review]

**Pull request merged**